### PR TITLE
added exceptional handling in user

### PR DESCRIPTION
**Pull request by pulkitdhingraoyo**
Code Reviewer Checklist:
- [ ] Tech Documentation : On Knowledge/ Google Doc
- [ ] Proper formatting, indentation, No spelling mistakes
- [ ] Meaningful Variable, Function, Class Names etc.
- [ ] Proper usage of access modifiers - private, public etc
- [ ] No Duplicate Code
- [ ] PR should be less than 10 files
- [ ] Classes should be less than 200 lines
- [ ] Functions should be less than 20 lines
- [ ] Code Documentation
- [ ] Test Cases
- [ ] Local Functions (Function Group) optimized for space, time & resources used
- [ ] No unnecessary and unused variables
- [ ] No unnecessary database calls in code (Cache usage if possible)
- [ ] Proper Error Handling in place
- [ ] Proper Logging (such a way that it is easier to find it on Kibana)

### CODE REVIEW

File: src/main/java/com/example/hbs/exception/ExceptionHandler.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
+package com.example.hbs.exception;
```
Comment (harisoyo) [inline src/main/java/com/example/hbs/exception/ExceptionHandler.java]: @aryan-dua already did this

File: src/main/java/com/example/hbs/exception/HbsException.java [hunk 2]
```diff
@@ -0,0 +1,25 @@
+package com.example.hbs.exception;
```
Comment (harisoyo) [inline src/main/java/com/example/hbs/exception/HbsException.java:1]: same as above

File: src/main/java/com/example/hbs/service/UserService.java [hunk 3]
```diff
@@ -50,20 +52,21 @@ public UserResponseDto addUser(UserRequestDto userRequestDto) {
...
+        if (userRequestDto.getUserEmail() != null) {
+            user.setUserEmail(userRequestDto.getUserEmail());
+        }
+        if (userRequestDto.getUserContact() != null) {
+            user.setUserContact(userRequestDto.getUserContact());
+        }
+        if (userRequestDto.getUserRole() != null) {
+            user.setUserRole(userRequestDto.getUserRole());
+        }
+        return userMapper.map(userRepository.save(user));
```
Comment (harisoyo) [inline src/main/java/com/example/hbs/service/UserService.java]: we should throw an exception if some required value is missing
like role, email etc.

**Review by harisoyo — CHANGES_REQUESTED** [review]

File: src/main/java/com/example/hbs/exception/ExceptionHandler.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.ControllerAdvice;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+@ControllerAdvice
+public class ExceptionHandler {
```
Comment (vivekSinha-OYO) [inline src/main/java/com/example/hbs/exception/ExceptionHandler.java]: name it to HbsExceptionHandler

File: src/main/java/com/example/hbs/service/UserService.java [hunk 5]
```diff
@@ -50,20 +66,21 @@ public UserResponseDto addUser(UserRequestDto userRequestDto) {
...
-            if (userRequestDto.getUserEmail() != null)
-                user.setUserEmail(userRequestDto.getUserEmail());
-            if (userRequestDto.getUserContact() != null)
-                user.setUserContact(userRequestDto.getUserContact());
-            if (userRequestDto.getUserRole() != null)
-                user.setUserRole(userRequestDto.getUserRole());
-            return userMapper.map(userRepository.save(user));
-        } else {
-            return null;
+        User user = userRepository.findById(id).orElse(null);
```
Comment (vivekSinha-OYO) [inline src/main/java/com/example/hbs/service/UserService.java]: you can throw exception here only,
`User user = userRepository.findById(id).orElseThrow(() -> new HbsException())`

File: src/main/java/com/example/hbs/service/UserService.java [hunk 6]
```diff
@@ -24,6 +26,20 @@
     private UserMapper userMapper;
 
     public UserResponseDto addUser(UserRequestDto userRequestDto) {
+
+        if (userRequestDto.getUserName() == null) {
```
Comment (vivekSinha-OYO) [inline src/main/java/com/example/hbs/service/UserService.java]: check for both null and empty string

File: src/main/java/com/example/hbs/service/UserService.java [hunk 7]
```diff
@@ -24,6 +26,20 @@
...
+            throw new HbsException("UserName is missing");
+        }
+        if (userRequestDto.getUserEmail() == null) {
```
Comment (vivekSinha-OYO) [inline src/main/java/com/example/hbs/service/UserService.java]: similarly, for the rest

File: src/main/java/com/example/hbs/exception/ExceptionHandler.java [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+
+    @org.springframework.web.bind.annotation.ExceptionHandler({HbsException.class})
+    public ResponseEntity<Object> handleException(
+            HbsException hotelException) {
+        List<String> exception = new ArrayList<>(Collections.singletonList(hotelException.getMessage()));
+        ResponseDto<Object> responseDto = new ResponseDto<>(exception);
+        return new ResponseEntity<>(responseDto, HttpStatus.BAD_REQUEST);
+    }
```
Comment (vivekSinha-OYO) [inline src/main/java/com/example/hbs/exception/ExceptionHandler.java]: only one kind kf exception??
what if you have other exception and with some different http status code
like in many cases we return UNPROCESSABLE_ENTITY

File: src/main/java/com/example/hbs/exception/HbsException.java [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+
+import lombok.Getter;
+import org.springframework.http.HttpStatus;
+
+@Getter
+public class HbsException extends RuntimeException{
```
Comment (vivekSinha-OYO) [inline src/main/java/com/example/hbs/exception/HbsException.java:7]: other exceptions should inherit from it.

File: src/main/java/com/example/hbs/exception/HbsExceptionHandler.java [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+import org.springframework.web.bind.annotation.ControllerAdvice;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+@ControllerAdvice
+public class HbsExceptionHandler {
+
+    @org.springframework.web.bind.annotation.ExceptionHandler({HbsException.class})
```
Comment (vivekSinha-OYO) [inline src/main/java/com/example/hbs/exception/HbsExceptionHandler.java]: `@ExceptionHandler`
and import the interface

File: src/main/java/com/example/hbs/service/UserService.java [hunk 11]
```diff
@@ -50,20 +67,21 @@ public UserResponseDto addUser(UserRequestDto userRequestDto) {
...
-                user.setUserEmail(userRequestDto.getUserEmail());
-            if (userRequestDto.getUserContact() != null)
-                user.setUserContact(userRequestDto.getUserContact());
-            if (userRequestDto.getUserRole() != null)
-                user.setUserRole(userRequestDto.getUserRole());
-            return userMapper.map(userRepository.save(user));
-        } else {
-            return null;
+        User user = userRepository.findById(id).orElseThrow(() -> new HbsException("No such user id found"));
+        //Optional.ofNullable(user).orElseThrow(() -> new HbsException("No such user Id found"));
```
Comment (vivekSinha-OYO) [inline src/main/java/com/example/hbs/service/UserService.java]: remove unused code

File: src/main/java/com/example/hbs/exception/HbsException.java [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+    public HbsException(String message) {
+        super(message);
+    }
+
+    public HbsException(String message, Exception exception) {
+        super(message, exception);
+    }
+
+
+}
```
Comment (vivekSinha-OYO) [inline src/main/java/com/example/hbs/exception/HbsException.java]: a new line at the end

File: src/main/java/com/example/hbs/service/UserService.java [hunk 13]
```diff
@@ -24,6 +26,21 @@
...
+            throw new HbsException("UserName is missing");
+        }
+
+        if (userRequestDto.getUserEmail() == null || userRequestDto.getUserEmail().equals("")) {
+            throw new HbsException("User Email is missing");
+        }
+        if (userRequestDto.getUserContact() == null || userRequestDto.getUserContact().equals("")) {
+            throw new HbsException("User contact is missing");
+        }
+        if (userRequestDto.getUserRole() == null ) {
```
Comment (harisoyo) [inline src/main/java/com/example/hbs/service/UserService.java:40]: Objects.isNull

**Review by harisoyo — APPROVED** [review]

**Pull request merged**